sherpa/optmethods/tests/test_optmethods.py

```diff
@@ -397,7 +397,11 @@ def test_Shekel7(opt, npar=4):
 def test_Shekel10(opt, npar=4):
     tst_opt(opt, _tstoptfct.Shekel10, npar)
 
-@pytest.mark.parametrize("opt", [minim, montecarlo, neldermead])
+@pytest.mark.parametrize("opt", [pytest.param(minim, marks=pytest.mark.xfail),
+                                 pytest.param(montecarlo,
+                                              marks=pytest.mark.xfail),
+                                 pytest.param(neldermead,
+                                              marks=pytest.mark.xfail)])
```

Comment:
How come these tests now fail?

Reply:
As @DougBurke pointed out in issue #1011, the initial free parameters are not set. So I restored the ```ShekelModifiedInit``` function in its original form:

```
(sherpa) [dtn@devel12 sherpa]$ git diff sherpa/optmethods/tests/tstoptfct.hh
diff --git a/sherpa/optmethods/tests/tstoptfct.hh b/sherpa/optmethods/tests/tstoptfct.hh
index 3ee65983..7b91ea0d 100644
--- a/sherpa/optmethods/tests/tstoptfct.hh
+++ b/sherpa/optmethods/tests/tstoptfct.hh
@@ -2823,12 +2823,12 @@ namespace tstoptfct {
     else
       throw std::runtime_error( "npar for the Shekel func must either 2 or 5\n" );
     mfct = 0;
-    srand(1);
+    // srand(1);
     for ( int ii = 0; ii < npar; ++ii ) {
       lo[ ii ] = 0.0;
       hi[ ii ] = 10.0;
-      double tmp = rand( ) / (RAND_MAX + 1.0);
-      x[ ii ] = lo[ ii ] + ( hi[ ii ] - lo[ ii ] ) * tmp;
+      // double tmp = rand( ) / (RAND_MAX + 1.0);
+      // x[ ii ] = lo[ ii ] + ( hi[ ii ] - lo[ ii ] ) * tmp;
     }
   }
```

That is by commenting out the call to ```srand(1)``` and setting the initial values for the free parameters: 
```
+      // double tmp = rand( ) / (RAND_MAX + 1.0);
+      // x[ ii ] = lo[ ii ] + ( hi[ ii ] - lo[ ii ] ) * tmp;
```

Moreover, I enabled the test script to stop at the ```ShekelModified``` function:
```
(sherpa) [dtn@devel12 sherpa]$ git diff sherpa/optmethods/tests/test_optmethods.py
diff --git a/sherpa/optmethods/tests/test_optmethods.py b/sherpa/optmethods/tests/test_optmethods.py
index a340da3a..7a58f9cc 100644
--- a/sherpa/optmethods/tests/test_optmethods.py
+++ b/sherpa/optmethods/tests/test_optmethods.py
@@ -403,6 +403,8 @@ def test_Shekel10(opt, npar=4):
                                  pytest.param(neldermead,
                                               marks=pytest.mark.xfail)])
 def test_ShekelModified(opt, npar=2):
+    import pdb
+    pdb.set_trace()
     tst_opt(opt, _tstoptfct.ShekelModified, npar)
 
 @pytest.mark.parametrize("opt", [pytest.param(minim, marks=pytest.mark.xfail),
```
As one can see from the following debugging session:

```
(sherpa) [dtn@devel12 sherpa]$ pytest --pdb sherpa/optmethods/tests/test_optmethods.py 
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-4.6.2, py-1.8.0, pluggy-0.12.0
rootdir: /export/sherpa, inifile: pytest.ini
plugins: arraydiff-0.3, xdist-1.29.0, forked-1.0.2, openfiles-0.3.2, doctestplus-0.3.0, remotedata-0.3.1
collected 233 items                                                            

....

> /export/sherpa/sherpa/optmethods/tests/test_optmethods.py(408)test_ShekelModified()
-> tst_opt(opt, _tstoptfct.ShekelModified, npar)
(Pdb) s
--Call--
> /export/sherpa/sherpa/optmethods/tests/test_optmethods.py(31)tst_opt()
-> def tst_opt(opt, fct, npar, reltol=1.0e-3, abstol=1.0e-3):
(Pdb) n
> /export/sherpa/sherpa/optmethods/tests/test_optmethods.py(41)tst_opt()
-> x0, xmin, xmax, fmin = init(fct.__name__, npar)
(Pdb) n
> /export/sherpa/sherpa/optmethods/tests/test_optmethods.py(42)tst_opt()
-> status, x, fval, msg, xtra = opt(fct, x0, xmin, xmax)
(Pdb) print(x0, xmin, xmax)
[1000. 1000.] [0. 0.] [10. 10.]
(Pdb) n
> /export/sherpa/sherpa/optmethods/tests/test_optmethods.py(43)tst_opt()
-> assert fmin == pytest.approx(fval, rel=reltol, abs=abstol)
(Pdb) print(fmin, fval)
-12.11900837975 -12.119008379753586
```

The initial free parameters are now ```[1000. 1000.]``` whereas the min and max of the parameters are: ```[0. 0.]``` and ```[10. 10.]``` . respectively.  The test does pass in spite of the out of bounds free parameters.  All optimization routines get called the function ```_check_args```:

```
def _move_within_limits(x, xmin, xmax):
    below = numpy.flatnonzero(x < xmin)
    if below.size > 0:
        x[below] = xmin[below]

    above = numpy.flatnonzero(x > xmax)
    if above.size > 0:
        x[above] = xmax[above]

def _check_args(x0, xmin, xmax):
    x = numpy.array(x0, numpy.float_)  # Make a copy
    xmin = numpy.asarray(xmin, numpy.float_)
    xmax = numpy.asarray(xmax, numpy.float_)

    if (x.shape != xmin.shape) or (x.shape != xmax.shape):
        raise TypeError('input array sizes do not match')

    _move_within_limits(x, xmin, xmax)

    return x, xmin, xmax
```

which basically moves the free parameters to the limits if the free parameters are out of bounds.  My guess is that if one set the initial free parameters to [10. 10.] in the ```ShekelModifiedInit``` function then the test will pass but that would be inconsistent with the other init test functions.  Choose your poison!

Reply:
So it's just luck that the code was passing before (by luck I mean "10,10" is a magical place). Okay, I see now why you have it as failing.